### PR TITLE
[BEAM-9167] Reduce Go SDK metric overhead

### PR DESCRIPTION
**Pull request by lostluck**
This PR dramatically reduces the overhead of metrics in the Go SDK.

A contemporary side by side comparison of the benchmark in the metrics package on my current machine:
```
benchmark                                        old ns/op     new ns/op     delta
BenchmarkMetrics/counter_inplace-12              585           249           -57.44%
BenchmarkMetrics/distribution_inplace-12         622           270           -56.59%
BenchmarkMetrics/gauge_inplace-12                812           311           -61.70%
BenchmarkMetrics/counter_predeclared-12          227           15.8          -93.04%
BenchmarkMetrics/distribution_predeclared-12     282           24.0          -91.49%
BenchmarkMetrics/gauge_predeclared-12            389           63.7          -83.62%

benchmark                                        old allocs     new allocs     delta
BenchmarkMetrics/counter_inplace-12              4              1              -75.00%
BenchmarkMetrics/distribution_inplace-12         4              1              -75.00%
BenchmarkMetrics/gauge_inplace-12                4              1              -75.00%
BenchmarkMetrics/counter_predeclared-12          3              0              -100.00%
BenchmarkMetrics/distribution_predeclared-12     3              0              -100.00%
BenchmarkMetrics/gauge_predeclared-12            3              0              -100.00%

benchmark                                        old bytes     new bytes     delta
BenchmarkMetrics/counter_inplace-12              160           48            -70.00%
BenchmarkMetrics/distribution_inplace-12         192           48            -75.00%
BenchmarkMetrics/gauge_inplace-12                192           48            -75.00%
BenchmarkMetrics/counter_predeclared-12          48            0             -100.00%
BenchmarkMetrics/distribution_predeclared-12     80            0             -100.00%
BenchmarkMetrics/gauge_predeclared-12            80            0             -100.00%
```

In particular this PR moves away from a global datastore for all metrics towards a perBundle based countersets. This allows for the removal of the per layer locks and the global lock that needed to be checked since all bundles had to check the same datastore. Now they only store a metric cell in the global store on first creation (still stored per bundle and per ptransform).

A subsequent change will remove the global store altogether in favour of better exposing the metrics per bundle, and allowing a callback visitor to thread-safely access the data inside each metric. This will also permit removing the dependency on the protos from the package, which was a mistake I made when I first wrote the package.

Further, Counters now use atomic operations rather than locks, which additional speeds them up vs the previous mutex approach.

Counter "names" are hashed ahead of time and the hash value cached in the proxy to increase the speed of subsequent lookups using the same proxy object.

This does make the proxies unsafe to use concurrently within the same bundle prior to first use, but this matches the general rule of Beam runners managing the concurrency for efficient processing, and that framework constructs are not safe for concurrent use by user code, without user managed locks.

As an exploration, I did try using sync.Map to avoid the above restriction, but the overhead for the additional interface wraping and unwraping was significant enough that this approach was worthwhile. 
This may be worth revisiting if Go gains Generics, as that would probably avoid this cost.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @youngoli

File: sdks/go/pkg/beam/core/metrics/metrics.go [hunk 1]
```diff
@@ -64,22 +67,57 @@ import (
...
+	css []*ptCounterSet
+}
+
+type nameHash uint64
+
+// ptCounterSet is the internal tracking struct for a single ptransform
+// in a single bundle for all counter types.
+type ptCounterSet struct {
+	// We store the user path access to the cells in metric type segregated
+	// maps. At present, caching the name hash,
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/core/metrics/metrics.go]: Thanks! This is a sentence fragment. I'll fix it in the next PR which needs to update this code a bit anyway.

File: sdks/go/pkg/beam/core/metrics/metrics.go [hunk 2]
```diff
@@ -64,22 +67,57 @@ import (
...
 type beamCtx struct {
 	context.Context
 	bundleID, ptransformID string
+	bs                     *perBundle
+	cs                     *ptCounterSet
 }
 
-// Value lifts the beam contLift the keys value for faster lookups when not available.
+// Value implements context.Value for beamCtx, and lifts the
+// values for metrics keys for value for faster lookups.
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/core/metrics/metrics.go]: That last "for value" is definitely out of place. 
In the next PR, I'll replace it with the following:
 Value implements the context.Value interface method for beamCtx. The implementation lifts the
 stored values for metrics keys to the top level beamCtx for faster lookups.

[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/metrics/metrics.go]: That works really well, sounds good

File: sdks/go/pkg/beam/core/metrics/metrics.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/pkg/beam/core/metrics/metrics.go]: sgtm

**Comment by lostluck** [comment]
So per the beam model, Bundles are the unit of a reproducible computation. If the bundle needs to be re-tried, the runner needs to be able to throw away the metrics for the bad bundle, so the final counters can be correct. As such, they are "logical" counters, rather than "physical" counters.

It's the runners job to aggregate metrics for all bundles, for final presentation, since any given worker isn't aware of the full distributed value of a counter.

To answer your questions: Yes, each bundle has their own instances of counters. These metrics get reported to the runner, and are currently "extracted" from the metric cells via the ToProto function (which requires the bundleID and PTransformID in question). It's the up to the runner to provide a way for the user to see them.
There's unfinished work on the "Job Management" proto to provide a common portable way for SDKs to get metrics from runners. Because of this, the Go SDK doesn't provide programmatic access to metrics after a job, so it's up to the user to access the data from the runner directly.

**Pull request merged**